src/coverage/coverageService.ts

```diff
@@ -0,0 +1,195 @@
+// Copyright 2019 Espressif Systems (Shanghai) CO LTD
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { sep, join } from "path";
+import * as vscode from "vscode";
+import { appendIdfAndToolsToPath, spawn } from "../utils";
+import { OutputChannel } from "../logger/outputChannel";
+import { Logger } from "../logger/logger";
+import * as idfConf from "../idfConfiguration";
+
+const COVERED_FACTOR = 0.9;
+const PARTIAL_FACTOR = 0.75;
+const MIN_LINE_COUNT = 1;
+
+export interface countRange {
+  range: vscode.Range;
+  count: Number;
+}
+
+export interface textEditorWithCoverage {
+  allLines: vscode.Range[];
+  countPerLines: countRange[];
+  coveredLines: vscode.Range[];
+  editor: vscode.TextEditor;
+  partialLines: vscode.Range[];
+  uncoveredLines: vscode.Range[];
+}
+
+export async function buildJson(dirPath: string) {
+  const idfPathDir =
+    idfConf.readParameter("idf.espIdfPath") || process.env.IDF_PATH;
+  const componentsDir = join(idfPathDir, "components");
+  const result = await _runCmd(
+    "gcovr",
+    [
+      "--filter",
+      ".",
+      "--filter",
+      componentsDir,
+      "--gcov-executable",
+      "xtensa-esp32-elf-gcov",
+      "--json",
+    ],
+    dirPath
+  );
+  return JSON.parse(result);
+}
+
+export async function buildHtml(dirPath: string) {
+  const idfPathDir =
+    idfConf.readParameter("idf.espIdfPath") || process.env.IDF_PATH;
+  const componentsDir = join(idfPathDir, "components");
+  const result = await _runCmd(
+    "gcovr",
+    [
+      "--filter",
+      ".",
+      "--filter",
+      componentsDir,
+      "--gcov-executable",
+      "xtensa-esp32-elf-gcov",
+      "--html",
+    ],
+    dirPath
+  );
+  return result;
+}
+
+function _runCmd(cmd: string, args: string[], dirPath: string) {
+  const modifiedEnv = appendIdfAndToolsToPath();
+  return spawn(cmd, args, { env: modifiedEnv, cwd: dirPath })
+    .then((resultBuffer) => resultBuffer.toString())
+    .catch((e) => {
+      const msg = e.error ? e.error.message : e;
+      Logger.error("Error on gcov cmd.\n" + msg, e);
+      OutputChannel.appendLine("Error building gcov cmd.\n" + msg);
+      return "";
+    });
+}
+
+export async function generateCoverageForEditors(
+  editors: vscode.TextEditor[],
+  dirPath: string
+) {
+  const gcovObj = await buildJson(dirPath);
```

Comment:
For every file opened we still calling build json and sometimes (getting error) (_attached screenshot_)

```bash
rejected promise not handled within 1 second: SyntaxError: Unexpected end of JSON input
extensionHostProcess.js:957
stack trace: SyntaxError: Unexpected end of JSON input
	at JSON.parse (<anonymous>)
	at /Volumes/Data/source-code/vscode-esp-idf-extension-gh/dist/extension.js:40434:21
	at Generator.next (<anonymous>)
	at fulfilled (/Volumes/Data/source-code/vscode-esp-idf-extension-gh/dist/extension.js:40406:58)
	at runMicrotasks (<anonymous>)
	at processTicksAndRejections (internal/process/task_queues.js:85:5)
```

<img width="1167" alt="Screen Shot 2020-05-21 at 1 00 24 PM" src="https://user-images.githubusercontent.com/25585652/82534618-0c616b00-9b63-11ea-9a79-172c778258ae.png">

Reply:
moved the buildJson call to the renderer object hoping to reduce the amount of times being called.